includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -717,11 +718,37 @@ public function remove_all_actions_on_preflight_check( $response, $handler, $req
 
 			// Avoid creating new accounts during preflight check.
 			remove_all_filters( 'woocommerce_checkout_registration_required' );
+
+			add_filter( 'woocommerce_coupon_get_usage_limit', '__return_null' );
+		} elseif (
+			WooPay_Session::is_request_from_woopay() &&
+			$this->is_request_to_store_api_checkout() &&
+			WooPay_Session::has_valid_request_signature()
+		) {
+			// We disable the user usage limit check during WooPay actual checkout because
+			// the coupon validation looks for existing orders with the same user email and coupon code,
+			// and the draft order generated during the preflight check meets those criteria.
+			// There is no problem in disabling it because the user usage limit is already checked
+			// during the preflight check.
+			add_filter( 'woocommerce_coupon_get_usage_limit_per_user', '__return_zero' );
```

Comment:
Why are we skipping the check during the actual checkout api call instead of during the pre-flight? 

I simulated a pre-flight fail and then tried to place the order again. The coupon got removed because it was already used.

Moving the filter inside the if made it work as expected.